### PR TITLE
[xbuild] Manage host/build contexts inside the graph

### PR DESCRIPTION
**Pull request by jgsogo**
Changelog: (Feature | Fix | Bugfix): Describe here your pull request
Docs: https://github.com/conan-io/docs/pull/XXXX

This PR is here to serve as an overview of the current implementation of the cross building model. Feel free to comment about it, but do not review in detail as many things will change (and probably this one won't be the final PR).

This is implementing **only the _backend_** part of the feature, for the command line interface, visit https://github.com/jgsogo/conan/pull/13.

There are some changes from this PR that are on its own PR to review them better or to be merged in advance:
 * Helper class to store `Node` objects: https://github.com/conan-io/conan/pull/5179

### CODE REVIEW

File: conans/client/graph/graph_builder.py [hunk 1]
```diff
@@ -54,20 +53,24 @@ def extend_build_requires(self, graph, node, build_requires_refs, check_updates,
 
         conanfile = node.conanfile
         scope = conanfile.display_name
-        requires = [Requirement(ref) for ref in build_requires_refs]
+        requires = []
+        for ref, context in build_requires_refs:
+            r = Requirement(ref)
```
Comment (SSE4) [inline conans/client/graph/graph_builder.py]: may be add ctor which has both ref and context?

Comment (jgsogo) [reply conans/client/graph/graph_builder.py]: Please, do not enter into details yet, there will be plenty of time to review the PR 😉  Better start with the look-and-feel of the tests (and take into account that this is under heavy development and I might change the implementation at any moment)

**Comment by jgsogo** [comment]
Refer to this PR instead: https://github.com/conan-io/conan/pull/5202

**Comment by jgsogo** [comment]
I'm adding some changes given the feedback in #5202, and preparing for the PR:
 * Default context for a `build_requires` is `host` (needed to keep existing behavior)
 * Remove some unnecessary changes: minor changes in tests, format or comments
 * Merged changes from `develop`, trying to keep track of the modifications introduced in the graph: comments, some new functions, and some lines removed. Changes introduced here, https://github.com/conan-io/conan/pull/5298, please double check.

**Comment by jgsogo** [comment]
Superseded by https://github.com/conan-io/conan/pull/5592

**Pull request closed without merging**